File: src/api/question_pool_stewardship.ts
// Closed Pool endorsement and actor-private subscription contracts.

import type { QuestionId } from "../../generated/api/QuestionId";

/** One immutable vetted public name, never an Account or Profile reference. */
export interface QuestionPoolStarredInstructor {
  readonly displayName: string;
}

/** Public endorsement facts available only to an authorized vetted Instructor. */
export interface QuestionPoolStarProjection {
  readonly starCount: number;
  readonly viewerHasStarred: boolean;
  readonly starredInstructors: ReadonlyArray<QuestionPoolStarredInstructor>;
}

/** The current Instructor's state only; no watcher list, count or notification. */
export interface QuestionPoolWatchProjection {
  readonly watching: boolean;
}

/** Explicit own-state operations for one stable public Pool identity. */
export interface QuestionPoolStewardshipClient {
  readonly getQuestionPoolStar: (poolId: QuestionId) => Promise<QuestionPoolStarProjection>;
  readonly setQuestionPoolStar: (
    poolId: QuestionId,
    starred: boolean,
  ) => Promise<QuestionPoolStarProjection>;
  readonly getQuestionPoolWatch: (poolId: QuestionId) => Promise<QuestionPoolWatchProjection>;
  readonly setQuestionPoolWatch: (
    poolId: QuestionId,
    watching: boolean,
  ) => Promise<QuestionPoolWatchProjection>;
}

File: crates/learning-data-access/src/question_pool_stewardship.rs
//! Stable published Pool Stars and actor-private Watches.
use crate::{SessionTokenHash, StoreError};
use async_trait::async_trait;
use question_model::QuestionId;

/// Approved public endorsement facts, with no Account/Profile or Watch fields.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct QuestionPoolStarProjection {
    pub viewer_has_starred: bool,
    pub star_count: u64,
    pub starred_instructors: Vec<QuestionPoolStarredInstructor>,
}

/// The immutable vetted public name, not an Account identifier.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct QuestionPoolStarredInstructor {
    pub display_name: String,
}

/// Subscription state for the authenticated actor only.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct QuestionPoolWatchProjection {
    pub watching: bool,
}

/// Database-authorized lineage relationships; retries set explicit state.
#[async_trait]
pub trait QuestionPoolStewardshipStore: Send + Sync {
    async fn question_pool_star_projection(
        &self,
        session_token_hash: SessionTokenHash,
        question_pool_id: &QuestionId,
    ) -> Result<QuestionPoolStarProjection, StoreError>;
    async fn set_current_question_pool_star_projection(
        &self,
        session_token_hash: SessionTokenHash,
        question_pool_id: &QuestionId,
        starred: bool,
    ) -> Result<QuestionPoolStarProjection, StoreError>;
    async fn question_pool_watch_projection(
        &self,
        session_token_hash: SessionTokenHash,
        question_pool_id: &QuestionId,
    ) -> Result<QuestionPoolWatchProjection, StoreError>;
    async fn set_current_question_pool_watch_projection(
        &self,
        session_token_hash: SessionTokenHash,
        question_pool_id: &QuestionId,
        watching: bool,
    ) -> Result<QuestionPoolWatchProjection, StoreError>;
}
